source/bigint.c

```diff
@@ -448,254 +456,287 @@ int aws_bigint_bytebuf_debug_output(const struct aws_bigint *bigint, struct aws_
-        } else if (lhs_digit > rhs_digit) {
-            return AWS_BI_GREATER_THAN;
-        }
+        /* does an assignment iff both conditions are true */
+        most_significant_digit_index ^= (mask_combined & index);
     }
 
-    return AWS_BI_EQUAL_TO;
+    return most_significant_digit_index;
 }
 
-bool aws_bigint_equals(const struct aws_bigint *lhs, const struct aws_bigint *rhs) {
-    AWS_PRECONDITION(aws_bigint_is_valid(lhs));
-    AWS_PRECONDITION(aws_bigint_is_valid(rhs));
+/*
+ * In the following functions gt and eq are updated in little "blocks".  After each block update, the variables will be
+ * in one of the following states:
+ *
+ *  (1) gt is 0, eq is 1, and from an ordering perspective, lhs == rhs, as checked "so far"
+ *  (2) gt is 1, eq is 0, (lhs > rhs)
+ *  (3) gt is 0, eq is 0, (lhs < rhs)
+ *
+ *  States (2) and (3) are terminal states that cannot be exited since eq is 0 and is the and-wise mask of all
+ *  subsequent gt updates.  Similarly, once eq is zero it cannot ever become non-zero.
+ *
+ *  Intuitively these ideas match the standard way of comparing magnitude equality by considering digit count and
+ *  digits from most significant to least significant.
+ *
+ *  Let l and r be the the two "components" (a digit, the number of significant digits) that we are
+ *  comparing between lhs and rhs.  Assume l and r are both non-negative (or +/- 1) and can each be represented
+ *  by an int64:
+ *
+ *  gt is maintained by the following bit trick:
+ *
+ *      l > r <=>
+ *      (r - l) < 0 <=>
+ *      (r - l) as an int64 has the high bit set <=>
+ *      ((r - l) >> 63) & 0x01 == 1
+ *
+ *  eq is maintained by the following bit trick:
+ *
+ *      l == r <=>
+ *      l ^ r == 0 <=>
+ *      (l ^ r) - 1 == -1 <=>
+ *      (((l ^ r) - 1) >> 63) & 0x01 == 1   // only true if l and r are < (1U << 63)
+ *
+ *      I found this last step confusing and a little uncomfortable.  Everywhere else we naturally think of l and
+ *      r as arbitrary, but here there's a bound as to under what conditions that last equivalence holds.  Having
+ *      a bigint with 1U << 63 or more digits is going to be an OOM error, but the failure of the equivalence
+ *      under that condition was a real sticking point to my initial understanding.
+ *
+ */
 
-    return s_aws_bigint_get_magnitude_ordering(lhs, rhs) == AWS_BI_EQUAL_TO && lhs->sign == rhs->sign;
-}
+static void s_aws_bigint_compare_magnitudes_helper(
+    const struct aws_bigint *lhs,
+    const struct aws_bigint *rhs,
+    volatile uint8_t *gt_flag,
+    volatile uint8_t *eq_flag) {
+    volatile uint8_t gt = *gt_flag;
+    volatile uint8_t eq = *eq_flag;
 
-bool aws_bigint_not_equals(const struct aws_bigint *lhs, const struct aws_bigint *rhs) {
-    AWS_PRECONDITION(aws_bigint_is_valid(lhs));
-    AWS_PRECONDITION(aws_bigint_is_valid(rhs));
+    /* most significant digit queries */
+    volatile int64_t lhs_msd = (int64_t)s_get_most_significant_digit_index(lhs);
```

Comment:
Isn't this const? Why would these vars need to be volatile?

Reply:
standard attempt at preventing optmization, new explanatory comment with #define